porcupine/plugins/anchors.py

```diff
@@ -0,0 +1,58 @@
+import tkinter
+import time
+from porcupine import tabs, get_tab_manager
+from porcupine.plugins import linenumbers
+
+# Dependent on code from linenumbers.py
+setup_after = ["linenumbers"]
+
+
+class AnchorManager:
+    def __init__(self, tab_textwidget: tkinter.Text, linenumber_instance: tkinter.Canvas) -> None:
+        self.tab_textwidget = tab_textwidget
+        self.linenumber_instance = linenumber_instance
+
+        self.anchor_symbol = "¶"
+
+        linenumber_instance.bind("<<Updated>>", self.update_linenumbers, add=True)
+
+    def toggle_on_off(self, event):
+        [cursor_index] = [self.tab_textwidget.index("insert linestart")]
+        for anchor in self.tab_textwidget.mark_names():
+            if anchor.startswith("anchor_") and self.tab_textwidget.index(anchor) == cursor_index:
+                self.tab_textwidget.mark_unset(anchor)
+                self.linenumber_instance.do_update()
+                return
+        self.tab_textwidget.mark_set("anchor_" + str(time.time()), "insert linestart")
+
+        self.linenumber_instance.do_update()
+
+    def update_linenumbers(self, event):
+        """Re-draws the anchor points every time the linenumber instance updates (scroll, insertion/deletion of text)"""
+        anchor_list = [
+            anchor for anchor in self.tab_textwidget.mark_names() if anchor.startswith("anchor_")
+        ]
+
+        for anchorpoint in anchor_list:
+            row_tag = "line_" + self.tab_textwidget.index(anchorpoint).split(".")[0]
+            [row_id] = self.linenumber_instance.find_withtag(row_tag)
```

Comment:
In a long file, the linenumbers plugin doesn't draw invisible lines. If you scroll so that an anchor mark is not visible, `find_withtag` will return an empty tuple and this code will error. Change it to handle `find_withtag` returning an empty tuple.